GradientDescent/gd.cpp

```diff
@@ -0,0 +1,75 @@
+#include "..\testfun.hpp"
+#include<iostream>
+#include<Eigen/Dense>
+
+// Use template to allow any type to be passed to gd
+template<typename ObjFunc>
+Eigen::VectorXd gd(ObjFunc f, const int max_iter, const Eigen::VectorXd& int_val, const double rel_sol_change_tol,
+                const double grad_change_tol, const double gam){
+     // Learning rate is fixed value at the moment. Could change this?
+    // Intialise the step, which will be x_i and step_prev which will be x_{i-1} at...
+    //.. each step of the gd algorithm
+    Eigen::VectorXd step;
+    Eigen::VectorXd step_prev = int_val;
+    std::cout<<step_prev<<std::endl;
+
+    //  Init df and df_prev which are \nabla F(x_i) and \nabla F(x_{i-1}) respectively.
+    const int dim = int_val.size();
+    Eigen::VectorXd df_prev(dim);
+    Eigen::VectorXd df(dim);
+
+    // We need to fill df_prev for the algorithm to work. Set every element equal to 100 (any big number will do)..
+    df_prev.fill(100);
+
+    // ... the same with the relative change in the solution and the change in the gradient...
+    // ... for x_{i-1} to x_i
+    double rel_sol_change = 100;
+    double grad_change = 100;
+
+    int iter = 0;
+
+    // Whilst the max number of iterations has not been reached, the relative change in the solutions and ...
+    // ... the change in the gradient are both less than the prescribed tol we look...
+    // ... hopefully getting closer and closer to the true minimiser.
+    while(iter < max_iter && rel_sol_change > rel_sol_change_tol && grad_change > grad_change_tol){
+        df = f.gradient(step_prev);                                                     // Calculate \nabla F(x_{i-1})...
+        step = step_prev - gam*df;
+        rel_sol_change = abs(f.evaluate(step_prev) - f.evaluate(step)/f.evaluate(step));    // The relative change in solution x_i and x_{i-1}
+        // L2 norm of change in gradient from solution x_{i-1} to x_i
+        grad_change =  ((df - df_prev).norm())/df.norm() ;
```

Comment:
Actually this is my fault. Should we be using the stopping criteria as the norm of the gradient at a single step \| \nabla f (x_i)\| rather than the difference between gradients  \| \nabla f (x_i) -  \nabla f(x_{i-1})\|? Or is the one we're using OK?

Reply:
I have the same issue, I think the norm of the gradient might be preferable since this will be small near critical points - though this means the algorithm could potentially terminate near local maxima as well?

Reply:
Without checking the Hessian I don't think we can know if it's a max or a min. I think just checking the norm of the gradient will be sufficient -- if we get super unlucky and manage to stop on top of a local maximum we can work it out when we get there. Perhaps we could say the solution has to not change *and* the gradient be small? Then for any gradient based method you'll immediately make a large jump away?